Coverage

sullof opened the pull request:
__Test coverage__

`yarn run test-src-web`
runs tests against the source code and generates a coverage. The files are transpiled at runtime and executed it in a Chrome Headless environment.

`yarn run test-dist-node`
runs test against the transpiled code.

--------------------
_Original description (to understand the comments)_

This adds coverage with reports to see what is missed.
It uses Karma and is executing the tests in a Chrome Headless instance. So, in the future, we should be able to perform web tests (React, etc...).
I could handle some transformations. So, I moved statements like
```
import providers from 'lib/providers';
```
to
```
import providers from './lib/providers';
```
Also, I added a 
```
} catch (err) {
```
to where it was 
```
} catch {
```
Finally, I also removed
```
static providers = providers
```
and added at the end
```
TronWeb.providers = providers
```
In a future version we can reuse those experimental statements (I like them).

The tests must mirror the structure of the `src` folder and have the suffix `_test`. For example, to test `src/lib/trx.js` we have to add the file `test/lib/trx_test.js`.

sullof commented:
I also changed 
```
feeLimit = 1_000_000_000,
```
to
```
feeLimit = 1000000000,
```
to make it compatible with the `istanbul-instrumenter-loader`.

sullof commented:
@Kondax @kookiekrak Before merging it, please, verify that it works. Unfortunately here in the coworking I am still waiting for the net admin to allow me to access the not-standard https ports (like 8090, 8091, etc.), so I cannot verify if the example works.

Kondax commented:
Is there any way to use Babel with karma so we don't lose what you've mentioned above?

sullof commented:
I tried but it didn't work. I can give it another try.

sullof commented:
So, I tried to use `babel-loader` first and `istanbul-instrumenter-loader` post. But the coverage is not generated correctly and all the files are empty. I think that `istanbul-instrumenter-loader` uses `babel-loader` under the surface, but uses the version `^7.1.2` instead of `^8.0.2` we uses. I think than soon they will update it and we will be able to reverse the changes.

sullof commented:
I tried before to compile for node, but I was unable to find a way to have a coverage and I ended up using Karma, which works with a virtual browser.
I can go back to my previous attempt and try to make it working.

Kondax commented:
If possible we should target node as they're separate browsers (and
separate builds).

We should also test the built files along with the source files - they can
act differently if they're transpiled.
On Tue, Sep 18, 2018 at 12:51 PM Francesco Sullo <notifications@github.com>
wrote:

> I tried before to compile for node, but I was unable to find a way to have
> a coverage and I ended up using Karma, which works with a virtual browser.
> I can go back to my previous attempt and try to make it working.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tronprotocol/tron-web/pull/17#issuecomment-422362581>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQFo-v9pJ-tZepmK80VsV6oayNFBjJplks5ucN5agaJpZM4Wtqlt>
> .
>

sullof commented:
I agree.
Testing the transpiled files is trivial and can be easily managed with a ENV variable, even if we cannot test everything there, because the components are not exposed in the dist files.

I am focused on the node coverage. Hope that I can make it soon.

Kondax commented:
I've had a look and it might be a better idea to use babel-plugin-istanbul
instead of instanbul-instrument-loader.

It's more up to date and supports Babel 7.
On Tue, Sep 18, 2018 at 1:53 PM Francesco Sullo <notifications@github.com>
wrote:

> I agree.
> Testing the transpiled files is trivial and can be easily managed with a
> ENV variable, even if we cannot test everything there, because the
> components are not exposed in the dist files.
>
> I am focused on the node coverage. Hope that I can make it soon.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tronprotocol/tron-web/pull/17#issuecomment-422380806>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQFo-u0zsXIZxHzI7s7R93ZoF2yTvm68ks5ucOzNgaJpZM4Wtqlt>
> .
>

sullof commented:
You were right. `babel-plugin-istanbul` supports all the statements.
Also, I readded the test of the dist-node files. The solution for now is not the best, but it is hard to share the files among the two tests because one is tested in node and the other is tested in the browser. So, env variables, conf files, etc. don't work the same way in the two environments. So, for now, I just duplicated the files. I will think a better way to avoid code repetitions in a future PR.

Kondax commented:
Hmm, so karma cannot run tests for both NodeJS and Chrome headless, only one? We might want to try and rectify this so it doesn't become a mess in the future.

sullof commented:
I tried to add node test without success in another branch. In this PR I am testing the dist files, i.e., the original test, just moving it in a subfolder to avoid conflicts.

Kondax commented on test/dist-node/index-spec.js:
```diff
@@ -239,6 +239,15 @@ describe('TronWeb Instance', () => {
             assert.equal(isConnected, false);
         });
+
+        it('should return false if the eventServer is null', async () => {
```
We shouldn't directly modify `tronWeb.eventServer` (and I'd like to disable this through a proxy object). Instead you should even create an instance without an event server specified or call `tronWeb.setEventServer()` which will default to `false`.

Kondax commented on test/src-web/src/index_test.js:
```diff
@@ -239,6 +239,15 @@ describe('TronWeb Instance', () => {
             assert.equal(isConnected, false);
         });
+
+        it('should return false if the eventServer is null', async () => {
```
Same issue as mentioned above

sullof replied:
Same response as above, it is not really useful, but decreases the coverage if not.

Kondax commented on test/dist-node/index-spec.js:
```diff
@@ -350,4 +359,16 @@ describe('TronWeb Instance', () => {
     });
+
+    describe('#currentProvider()', () => {
+        it('should return the current providers', () => {
```
Do we need to test this? It's an alias of a function so does not actually do anything itself.

sullof replied:
I agree it is not really useful, but I add to have a better coverage.

Kondax commented on test/src-web/src/index_test.js:
```diff
@@ -350,4 +359,16 @@ describe('TronWeb Instance', () => {
     });
+
+    describe('#currentProvider()', () => {
+        it('should return the current providers', () => {
```
Same as above

Kondax commented:
What are the conflicts? We should be able to run both tests with the same test files.

sullof commented:
The way you load TronWeb is very different in the two environments. In the browser the entire object has been transpiled previously by Webpack, in node the object is required a runtime. There is no way to put there a condition and decide what to load based on the environment. At least, I haven't found one. I will think about it.

Kondax commented:
I was just looking for the best approach for doing the above and I've come across https://github.com/karma-runner/grunt-karma. We'd provide a base config file, and then two specific configs which are merged with the base config (one for node, one for browsers).

sullof commented:
It looks interesting. I was thinking of building a very simple automation tool to compose the test files before running the tests. This is basically the same approach.

sullof commented:
Testing on the other browsers you mentioned is feasible, since there is a karma launcher for all of them.

sullof commented:
@Kondax 
I added tests for Firefox and Edge. They work only if you have the browser installed on your machine. In my case, since I have a macbook, I don't have Edge. But the task is completed anyway ignoring the missing browser.
I have also added an initial configuration for grunt-karma.
Tomorrow morning I will continue.

Kondax commented:
Sounds good. Since this is a public repo (assuming you / Tian have appropriate permission) do we want to benefit from adding Travis or another CI for automatically running tests on PRs?

kookiekrak commented:
Works for me, no time right now to set it up though

> On Sep 18, 2018, at 10:16 AM, Kondax <notifications@github.com> wrote:
> 
> Sounds good. Since this is a public repo (assuming you / Tian have appropriate permission) do we want to benefit from adding Travis or another CI for automatically running tests on PRs?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

sullof commented:
`grunt-karma` allows to create multiple Karma configurations, but Karma works only for tests in browsers, so we cannot use it to test the code against Node.

I am simply using karma for tests in browsers to generate a coverage report. Separately, we run the tests against Node, starting from the transpiled code.
For Node tests, we don't need coverage (since we are going to cover the same cases). 

About using Travis, a simpler solution, at least for now, is to force tests before pushing. I added a `pre-push` script in git to force successful tests before pushing a change. It is more common to use a `pre-commit`, but this does not allow you to commit temporary changes if the tests fail, and sometimes that is very useful. Anyway, if there are unwanted commits, we can rebase.

Also, since Mocha uses its own context, it is a best practice to not use arrows in tests, but use `function`, so that the right context can be kept. In our specific case, this is important because the test `#isEventServerConnected()` often fails in some browsers (mostly Firefox) due to a timeout.
So, I set a different timeout only for that case.

Finally, I have left, for now, the Grunt-karma config in case we decide to use it anyway.

I think that this is a good starting point. What do you think?

sullof commented:
I also added in pre-push a check to block pushing to master. Of course, someone can change these rules before pushing, and this is why it would be better to use Travis CI or something similar, but for now, it is good, at least, to avoid pushing to master by mistake.

Kondax commented:
Seems like a good compromise. Regarding blocking master commits maybe Tian
could disable this in the repository settings?

For the pre-push and pre-commit checks do they only run on *nix? We need to
support Windows too, we could simply use a node script instead of bash
script to check the tests.
On Wed, Sep 19, 2018 at 12:34 PM Francesco Sullo <notifications@github.com>
wrote:

> I also added in pre-push a check to block pushing to master. Of course,
> someone can change these rules before pushing, and this is why it would be
> better to use Travis CI or something similar, but for now, it is good, at
> least, to avoid pushing to master by mistake.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tronprotocol/tron-web/pull/17#issuecomment-422767060>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQFo-rgEFcIphn96MQSphyWzoABalzSJks5ucivDgaJpZM4Wtqlt>
> .
>

sullof commented:
I agree on disabling push to master in the settings. @kookiekrak?

About the pre-push, you are right, it can work on Windows 10 under some circumstances, but it won't work on previous versions. I suggest we fix this in a future PR, though, since I can't do it now because I am working on Tronbox.

Kondax commented:
I'll be able to fix it, I'll do so later.
On Wed, Sep 19, 2018 at 2:48 PM Francesco Sullo <notifications@github.com>
wrote:

> I agree on disabling push to master in the settings. @kookiekrak
> <https://github.com/kookiekrak>?
>
> About the pre-push, you are right, it can work on Windows 10 under some
> circumstances, but it won't work on previous versions. I suggest we fix
> this in a future PR, though, since I can't do it now because I am working
> on Tronbox.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tronprotocol/tron-web/pull/17#issuecomment-422810381>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQFo-po8iRhPRbgSpgFiWVJCOGUYwoKcks5uckshgaJpZM4Wtqlt>
> .
>

kookiekrak commented:
i'll disable, need to get admin access first

Kondax commented:
Changes made in #20 
Test suite now supports multi-os.
